debezium-connector-postgres/src/main/java/io/debezium/connector/postgresql/PostgresStreamingChangeEventSource.java

```diff
@@ -232,6 +232,34 @@ else if (message.getOperation() == Operation.COMMIT) {
                     }
                     maybeWarnAboutGrowingWalBacklog(true);
                 }
+                else if (message.getOperation() == Operation.LOGICAL_DECODING_MESSAGE) {
+                    if (connectorConfig.logicalDecodingMessageHandlingMode() != PostgresConnectorConfig.LogicalDecodingMessageHandlingMode.INCLUDE) {
+                        LOGGER.debug("Received logical decoding message {}", message);
+                        if (message.isLastEventForLsn()) {
+                            commitMessage(partition, offsetContext, lsn);
+                        }
+                        return;
+                    }
+
+                    // non-transactional messages will not have a commitTime or txId
+                    if (message.isLastEventForLsn()) {
+                        offsetContext.updateWalPosition(lsn, lastCompletelyProcessedLsn, null, null,
+                                taskContext.getSlotXmin(connection));
+                        commitMessage(partition, offsetContext, lsn);
```

Comment:
In the case of a non-transactional message, there aren't any BEGIN or COMMIT messages surrounding them, so it seemed appropriate to use the same logic that's used in the COMMIT block above.
https://github.com/debezium/debezium/blob/fe26649b4b5eca179fa5e0a2f2511e1efd49f8e1/debezium-connector-postgres/src/main/java/io/debezium/connector/postgresql/PostgresStreamingChangeEventSource.java#L229-L233

The seems to just update the offsetContext and dispatch a heartbeat message, but I can remove that call to `commitMessage` if my assumptions are off base here. 
https://github.com/debezium/debezium/blob/fe26649b4b5eca179fa5e0a2f2511e1efd49f8e1/debezium-connector-postgres/src/main/java/io/debezium/connector/postgresql/PostgresStreamingChangeEventSource.java#L351-L356

Reply:
In the case of a transactional logical message, `isLastEventForLsn` would be `false` since the assumption is that a transactional logical decoding message will eventually be followed by a `COMMIT` message which would trigger the `commitMessage` call in the above if statement. We should only enter this branch if the logical decoding message is not transactional. 

https://github.com/debezium/debezium/blob/8007cc3b053d8299ccbd68d8dbde0a259dc52900/debezium-connector-postgres/src/main/java/io/debezium/connector/postgresql/connection/LogicalDecodingMessage.java#L35-L38

Your suggestion of checking `message.getTransactionId()` does bring up something I haven't mentioned with this new` LogicalDecodingMessage` class though. I was not able to set the `transactionId` of this new message type to `null` in the case of non-transactional messages because the type is `long` instead of `Long`, so `getTransactionId()` can never be null. In the case of a non-transactional message, the `transactionId` is just the ID of the last `BEGIN` message that the decoder encountered. We're obfuscating all of this by just setting the `txID` to an empty string in the `SourceInfo`. Do you think we can get away with changing the type of `getTransactionId()` in the `ReplicationMessage` interface to `Long` in a minor version update? If so, I can make that update in this PR as well. That feels more correct to me.

https://github.com/debezium/debezium/blob/8007cc3b053d8299ccbd68d8dbde0a259dc52900/debezium-connector-postgres/src/main/java/io/debezium/connector/postgresql/connection/pgoutput/PgOutputMessageDecoder.java#L232-L235

https://github.com/debezium/debezium/blob/8007cc3b053d8299ccbd68d8dbde0a259dc52900/debezium-connector-postgres/src/main/java/io/debezium/connector/postgresql/connection/pgoutput/PgOutputMessageDecoder.java#L601-L607

Reply:
> In the case of a transactional logical message, isLastEventForLsn would be false since the assumption is [...]

Ah yes, got it now. Probably worth a clarifying comment along those lines on that if statement.

> Do you think we can get away with changing the type of getTransactionId() in the ReplicationMessage interface to Long in a minor version update?

Yes, that's no problem at all; all these  types are not meant for external consumption/implementation, so we take the liberty to modify them as we see fit (admittedly, there is no precise separation between public API/SPI of Debezium and internal parts (yet), but `ReplicationMessage` definitely is internal. I'd only suggest to change this to `OptionalLong` rather than `Long`, so to make make its semantics more explicit.